### ISSUE TITLE
Add a field for span data in generated types

### ISSUE THREAD

**Issue by ryo33** [jsinger67/parol#44]
```rust
struct Span(Range<usize>);

struct  ExprLiteralInteger {
    span: Span, // this
    integer: Token,
}
```

This is very useful for constructing High-level IR with span data.

 Not that I need it right now, but I would like to work on this when I have time.

**Comment by ryo33** [comment]
I want this feature for my [Desk-lang](https://github.com/Hihaheho/Desk).
Also, I've heard my friend, the author of another language, also needs span data if `parol` would be used.

**Comment by ryo33** [comment]
It would be great if the `Span` struct could be defined in `parol_runtime` and shared everywhere, including #43.

**Comment by jsinger67** [comment]
Currently you can extract the span information from the token itself.
I can provide you an example for this later.

**Comment by ryo33** [comment]
At first, I regret that I forgot to mention that I do not believe that all people need span information. So it feature would be toggled by auto-generation configuration.

> Maybe we can resort to some kind of lazy evaluation of the span of an element that the user can initiate only if he needs this information. Best would be that the lazy evaluation methods are generated somehow automatically.

the lazy evaluation of span seems nice.

> But your argument, that a symbol clipped with ^ can't provide any span information is convincing too.
We should carefully think about the consequences,

Even if a terminal or non-terminal is clipped, we should include it in the generated type with `_` prefix of the field name.

**Comment by jsinger67** [comment]
I am convinced that we will find a satisfying solution for this issue.
Currently the span information is contained in the Token itself as a member of type Position.
There exist several possible solutions.
1. Use the position of the Token on client side in the desired way (current situation)
2. Provide an additional member *span* in the generated structures and fill it at compile time (enabled optionally)
3. Generate functions for each generated struct that calculate span information on demand.
4. ...

Further discussion is needed.

**Comment by jsinger67** [comment]
I'll start with the third option.
It should also be able to be enabled via command line option or per builder configuration.
I am currently assuming that we have all the necessary information to get `std::ops::Range` values pointing into the input string (as byte offsets!).
If a given generated type does not contain a token (and all other nonterminals recursively contain no terminals), it would return an empty Range value.
If the user needs scope information in a specific generated type that does not contain tokens recursively due to the cut operator (`^`) used, consider omitting the cut operator in this particular case. This will surely emerge as his grammar develops. Actually, range information is only necessary for error reporting or diagnostics. Clipping grammar symbols should therefore be possible in most cases.

**Comment by jsinger67** [comment]
## Fixed with commit e79e3dd

* `parol` can now optionally generate the `ToSpan` trait automatically for all generated AST types
* This can be enabled by `-r` CLI option or via builder configuration `range`

Please evaluated thoroughly!

Have fun! 🎉
